numba/core/typed_passes.py

```diff
@@ -956,3 +961,58 @@ def _strip_phi_nodes(self, func_ir):
 
         func_ir.blocks = newblocks
         return func_ir
+
+    def _simplify_conditionally_defined_variable(self, func_ir):
+        """
+        Rewrite assignments like:
+
+            ver1 = null()
+            ...
+            ver1 = ver
+            ...
+            uses(ver1)
+
+        into:
+            # delete all assignments to ver1
+            uses(ver)
+
+        This is only needed for parfors.
+        """
+        any_block = next(iter(func_ir.blocks.values()))
+        scope = any_block.scope
+        defs = func_ir._definitions
+
+        # Find and simplify conditionally defined variables
+        suspects = set()
+        for varname, deflist in defs.items():
+            unique_defs = set(deflist)
+            atleast_one_undef = any(map(lambda x: x == ir.UNDEFINED,
```

Comment:
Is this the same as `atleast_one_undef = ir.UNDEFINED in unique_defs`? Are there concerns about `ir.UNDEFINED` not being singleton under some condition?

Reply:
This part is now changed to https://github.com/numba/numba/pull/9255/files#diff-8da448c689273bcc37cdbc30f509dc9bc0a582921225308874101d173dd69eb3R994